File: backend/internal/api/router.go
package api

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"

	"github.com/ansible-talk/backend/internal/auth"
	"github.com/ansible-talk/backend/internal/config"
	"github.com/ansible-talk/backend/internal/contacts"
	"github.com/ansible-talk/backend/internal/crypto"
	"github.com/ansible-talk/backend/internal/messaging"
	"github.com/ansible-talk/backend/internal/stickers"
	"github.com/ansible-talk/backend/internal/storage"
)

// Server holds all dependencies for the API server
type Server struct {
	Config       *config.Config
	DB           *storage.PostgresDB
	Redis        *storage.RedisClient
	MinIO        *storage.MinIOClient
	AuthService  *auth.Service
	CryptoService *crypto.Service
	ContactsService *contacts.Service
	MessagingService *messaging.Service
	StickersService *stickers.Service
	WSHub        *Hub
}

// NewServer creates a new API server
func NewServer(cfg *config.Config, db *storage.PostgresDB, redis *storage.RedisClient, minio *storage.MinIOClient) *Server {
	authSvc := auth.NewService(db, redis, cfg)
	cryptoSvc := crypto.NewService(db)
	contactsSvc := contacts.NewService(db)
	messagingSvc := messaging.NewService(db, redis)
	stickersSvc := stickers.NewService(db, minio)

	return &Server{
		Config:          cfg,
		DB:              db,
		Redis:           redis,
		MinIO:           minio,
		AuthService:     authSvc,
		CryptoService:   cryptoSvc,
		ContactsService: contactsSvc,
		MessagingService: messagingSvc,
		StickersService: stickersSvc,
		WSHub:           NewHub(redis),
	}
}

// SetupRouter configures all API routes
func (s *Server) SetupRouter() *gin.Engine {
	if s.Config.Server.Environment == "production" {
		gin.SetMode(gin.ReleaseMode)
	}

	r := gin.New()
	r.Use(gin.Logger())
	r.Use(gin.Recovery())
	r.Use(corsMiddleware())

	// Health check
	r.GET("/health", s.healthCheck)

	// API v1
	v1 := r.Group("/api/v1")
	{
		// Auth routes (public)
		authRoutes := v1.Group("/auth")
		{
			authRoutes.POST("/otp/send", s.sendOTP)
			authRoutes.POST("/otp/verify", s.verifyOTP)
			authRoutes.POST("/register", s.register)
			authRoutes.POST("/login", s.login)
			authRoutes.POST("/refresh", s.refreshToken)
		}

		// Protected routes
		protected := v1.Group("")
		protected.Use(s.authMiddleware())
		{
			// User routes
			userRoutes := protected.Group("/users")
			{
				userRoutes.GET("/me", s.getCurrentUser)
				userRoutes.PUT("/me", s.updateCurrentUser)
				userRoutes.POST("/me/avatar", s.uploadAvatar)
				userRoutes.GET("/search", s.searchUsers)
			}

			// Auth management
			protected.POST("/auth/logout", s.logout)
			protected.POST("/auth/logout-all", s.logoutAll)

			// Device routes
			deviceRoutes := protected.Group("/devices")
			{
				deviceRoutes.GET("", s.getDevices)
				deviceRoutes.DELETE("/:id", s.removeDevice)
			}

			// Signal keys routes
			keyRoutes := protected.Group("/keys")
			{
				keyRoutes.POST("/register", s.registerKeys)
				keyRoutes.GET("/bundle/:user_id/:device_id", s.getKeyBundle)
				keyRoutes.GET("/count", s.getPreKeyCount)
				keyRoutes.POST("/prekeys", s.refreshPreKeys)
				keyRoutes.PUT("/signed-prekey", s.updateSignedPreKey)
			}

			// Contact routes
			contactRoutes := protected.Group("/contacts")
			{
				contactRoutes.GET("", s.getContacts)
				contactRoutes.POST("", s.addContact)
				contactRoutes.GET("/:id", s.getContact)
				contactRoutes.PUT("/:id", s.updateContact)
				contactRoutes.DELETE("/:id", s.deleteContact)
				contactRoutes.POST("/:id/block", s.blockContact)
				contactRoutes.POST("/:id/unblock", s.unblockContact)
				contactRoutes.GET("/blocked", s.getBlockedContacts)
				contactRoutes.POST("/sync", s.syncContacts)
			}

			// Conversation routes
			convRoutes := protected.Group("/conversations")
			{
				convRoutes.GET("", s.getConversations)
				convRoutes.POST("/direct", s.createDirectConversation)
				convRoutes.POST("/group", s.createGroupConversation)
				convRoutes.GET("/:id", s.getConversation)
				convRoutes.GET("/:id/messages", s.getMessages)
				convRoutes.POST("/:id/messages", s.sendMessage)
				convRoutes.POST("/:id/typing", s.sendTyping)
			}

			// Message routes
			msgRoutes := protected.Group("/messages")
			{
				msgRoutes.POST("/:id/delivered", s.markDelivered)
				msgRoutes.POST("/:id/read", s.markRead)
				msgRoutes.DELETE("/:id", s.deleteMessage)
			}

			// Sticker routes
			stickerRoutes := protected.Group("/stickers")
			{
				stickerRoutes.GET("/catalog", s.getStickerCatalog)
				stickerRoutes.GET("/search", s.searchStickers)
				stickerRoutes.GET("/packs/:id", s.getStickerPack)
				stickerRoutes.POST("/packs/:id/download", s.downloadStickerPack)
				stickerRoutes.DELETE("/packs/:id", s.removeStickerPack)
				stickerRoutes.GET("/my-packs", s.getUserStickerPacks)
				stickerRoutes.PUT("/my-packs/reorder", s.reorderStickerPacks)
			}

			// Admin sticker routes (for creating packs)
			adminStickerRoutes := protected.Group("/admin/stickers")
			{
				adminStickerRoutes.POST("/packs", s.createStickerPack)
				adminStickerRoutes.POST("/packs/:id/cover", s.uploadPackCover)
				adminStickerRoutes.POST("/packs/:id/stickers", s.addSticker)
			}
		}

		// WebSocket route
		v1.GET("/ws", s.authMiddleware(), s.handleWebSocket)
	}

	return r
}

// Health check handler
func (s *Server) healthCheck(c *gin.Context) {
	ctx := c.Request.Context()

	// Check database
	if err := s.DB.Health(ctx); err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{
			"status": "unhealthy",
			"error":  "database connection failed",
		})
		return
	}

	// Check Redis
	if err := s.Redis.Health(ctx); err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{
			"status": "unhealthy",
			"error":  "redis connection failed",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"status":    "healthy",
		"timestamp": time.Now().UTC(),
	})
}

// CORS middleware
func corsMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Writer.Header().Set("Access-Control-Allow-Origin", "*")
		c.Writer.Header().Set("Access-Control-Allow-Credentials", "true")
		c.Writer.Header().Set("Access-Control-Allow-Headers", "Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, Authorization, accept, origin, Cache-Control, X-Requested-With")
		c.Writer.Header().Set("Access-Control-Allow-Methods", "POST, OPTIONS, GET, PUT, DELETE, PATCH")

		if c.Request.Method == "OPTIONS" {
			c.AbortWithStatus(204)
			return
		}

		c.Next()
	}
}

// Auth middleware
func (s *Server) authMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		token := c.GetHeader("Authorization")
		if token == "" {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "authorization header required"})
			c.Abort()
			return
		}

		// Remove "Bearer " prefix if present
		if len(token) > 7 && token[:7] == "Bearer " {
			token = token[7:]
		}

		claims, err := s.AuthService.ValidateToken(c.Request.Context(), token)
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "invalid token"})
			c.Abort()
			return
		}

		// Set user info in context
		c.Set("user_id", claims.UserID)
		c.Set("device_id", claims.DeviceID)

		c.Next()
	}
}

// Helper to get user ID from context
func getUserID(c *gin.Context) string {
	userID, _ := c.Get("user_id")
	return userID.(string)
}

// Helper to get device ID from context
func getDeviceID(c *gin.Context) string {
	deviceID, _ := c.Get("device_id")
	return deviceID.(string)
}

File: backend-rs/src/api/router.rs
use axum::{
    middleware,
    routing::{delete, get, post, put},
    Router,
};

use super::{handlers, middleware::auth_middleware, websocket::handle_websocket};
use crate::AppState;

pub fn create_router(state: AppState) -> Router<AppState> {
    // Public auth routes
    let auth_routes = Router::new()
        .route("/otp/send", post(handlers::auth::send_otp))
        .route("/otp/verify", post(handlers::auth::verify_otp))
        .route("/register", post(handlers::auth::register))
        .route("/login", post(handlers::auth::login))
        .route("/refresh", post(handlers::auth::refresh_token));

    // Protected auth routes
    let auth_protected = Router::new()
        .route("/logout", post(handlers::auth::logout))
        .route("/logout-all", post(handlers::auth::logout_all))
        .layer(middleware::from_fn_with_state(state.clone(), auth_middleware));

    // User routes (protected)
    let user_routes = Router::new()
        .route("/me", get(handlers::users::get_current_user))
        .route("/me", put(handlers::users::update_current_user))
        .route("/me/avatar", post(handlers::users::upload_avatar))
        .route("/search", get(handlers::users::search_users))
        .layer(middleware::from_fn_with_state(state.clone(), auth_middleware));

    // Device routes (protected)
    let device_routes = Router::new()
        .route("/", get(handlers::devices::get_devices))
        .route("/:id", delete(handlers::devices::remove_device))
        .layer(middleware::from_fn_with_state(state.clone(), auth_middleware));

    // Key routes (protected)
    let key_routes = Router::new()
        .route("/register", post(handlers::keys::register_keys))
        .route("/bundle/:user_id/:device_id", get(handlers::keys::get_key_bundle))
        .route("/count", get(handlers::keys::get_pre_key_count))
        .route("/prekeys", post(handlers::keys::refresh_pre_keys))
        .route("/signed-prekey", put(handlers::keys::update_signed_pre_key))
        .layer(middleware::from_fn_with_state(state.clone(), auth_middleware));

    // Contact routes (protected)
    let contact_routes = Router::new()
        .route("/", get(handlers::contacts::get_contacts))
        .route("/", post(handlers::contacts::add_contact))
        .route("/:id", get(handlers::contacts::get_contact))
        .route("/:id", put(handlers::contacts::update_contact))
        .route("/:id", delete(handlers::contacts::delete_contact))
        .route("/:id/block", post(handlers::contacts::block_contact))
        .route("/:id/unblock", post(handlers::contacts::unblock_contact))
        .route("/blocked", get(handlers::contacts::get_blocked_contacts))
        .route("/sync", post(handlers::contacts::sync_contacts))
        .layer(middleware::from_fn_with_state(state.clone(), auth_middleware));

    // Conversation routes (protected)
    let conversation_routes = Router::new()
        .route("/", get(handlers::conversations::get_conversations))
        .route("/direct", post(handlers::conversations::create_direct_conversation))
        .route("/group", post(handlers::conversations::create_group_conversation))
        .route("/:id", get(handlers::conversations::get_conversation))
        .route("/:id/messages", get(handlers::conversations::get_messages))
        .route("/:id/messages", post(handlers::conversations::send_message))
        .route("/:id/typing", post(handlers::conversations::send_typing))
        .layer(middleware::from_fn_with_state(state.clone(), auth_middleware));

    // Message routes (protected)
    let message_routes = Router::new()
        .route("/:id/delivered", post(handlers::messages::mark_delivered))
        .route("/:id/read", post(handlers::messages::mark_read))
        .route("/:id", delete(handlers::messages::delete_message))
        .layer(middleware::from_fn_with_state(state.clone(), auth_middleware));

    // Sticker routes (public catalog, protected for user actions)
    let sticker_public_routes = Router::new()
        .route("/catalog", get(handlers::stickers::get_catalog))
        .route("/search", get(handlers::stickers::search_stickers))
        .route("/packs/:id", get(handlers::stickers::get_sticker_pack));

    let sticker_protected_routes = Router::new()
        .route("/packs/:id/download", post(handlers::stickers::download_sticker_pack))
        .route("/packs/:id", delete(handlers::stickers::remove_sticker_pack))
        .route("/my-packs", get(handlers::stickers::get_user_sticker_packs))
        .route("/my-packs/reorder", put(handlers::stickers::reorder_sticker_packs))
        .layer(middleware::from_fn_with_state(state.clone(), auth_middleware));

    // Admin sticker routes (protected - would need admin check in production)
    let admin_sticker_routes = Router::new()
        .route("/packs", post(handlers::stickers::create_sticker_pack))
        .route("/packs/:id/cover", post(handlers::stickers::upload_pack_cover))
        .route("/packs/:id/stickers", post(handlers::stickers::add_sticker))
        .layer(middleware::from_fn_with_state(state.clone(), auth_middleware));

    // WebSocket route (protected)
    let ws_route = Router::new()
        .route("/ws", get(handle_websocket))
        .layer(middleware::from_fn_with_state(state.clone(), auth_middleware));

    // Combine all routes
    Router::new()
        .nest("/auth", auth_routes.merge(auth_protected))
        .nest("/users", user_routes)
        .nest("/devices", device_routes)
        .nest("/keys", key_routes)
        .nest("/contacts", contact_routes)
        .nest("/conversations", conversation_routes)
        .nest("/messages", message_routes)
        .nest("/stickers", sticker_public_routes.merge(sticker_protected_routes))
        .nest("/admin/stickers", admin_sticker_routes)
        .merge(ws_route)
        .with_state(state)
}
